compiler/noirc_frontend/src/node_interner.rs

```diff
@@ -820,6 +820,23 @@ impl NodeInterner {
         self.id_to_type.get(&index.into()).cloned().unwrap_or(Type::Error)
     }
 
+    pub fn id_type_substitute_trait_as_type(&self, def_id: DefinitionId) -> Type {
+        let typ = self.id_type(def_id);
+        if let Type::Function(args, ret, env) = &typ {
+            let def = self.definition(def_id);
+            if let Type::TraitAsType(_trait) = ret.as_ref() {
+                if let DefinitionKind::Function(func_id) = def.kind {
+                    let f = self.function(&func_id);
+                    let func_body = f.as_expr();
+                    let ret_type = self.id_type(func_body);
+                    let new_type = Type::Function(args.clone(), Box::new(ret_type), env.clone());
+                    return new_type;
```

Comment:
This isn't quite correct since if the function advertises returning an `impl Display` but actually returns a `Foo`, this will lead to the caller seeing a `Foo` instead of something that is only known to implement Display. It's fine as a first pass though - we can save existential types for later.

Reply:
This is meant as first pass implementation, that will be refined later when we include more test cases.